### DISCUSSION TITLE
cross module dependency registration

### DISCUSSION THREAD

**Discussion by AndrewSB** [hmlongco/Factory#147] [Q&A]
i'm trying to get compile time safety for my dependencies in a multi module project. i can't figure out a way to compose that. here's an example

Module Interfaces
```swift
public protocol AuthClientType {
    func isLoggedIn() async -> Bool
    func token() async throws -> String?
}

extension Container {
    var authClient: Factory<AuthClientType> {
        fatalError()
    }
}
```

which is then used in Module FeatureA, that depends on Interfaces
```swift
import Interfaces 

struct V: View {
  @Injected(\.authClient) var authClient
}
```

then, in a top level module that imports everything, along with our 3rd party dependencies, i run 

```swift
import Interfaces
import Firebase
import BasicallyEveryDependency

func soonAfterAppLaunch() {
  Container.shared.authClient.register {
    FirebaseAuthClient()
  }
}
```

the crux of the issue is that both Interfaces & FeatureA are unaware of `FirebaseAuthClient` (by design, they don't import as many 3rd party concretions as possible).
i have a feeling that what i'm trying to do is impossible, but i'd love to be corrected

**Comment by AndrewSB** [comment]
i see https://github.com/hmlongco/Factory/issues/26, looks very similar. is this the suggested solution? all the dependencies should be made optional?

**Comment by AndrewSB** [comment]
the answer seems to be over here
https://github.com/hmlongco/Factory/issues/3#issuecomment-1167728263

i searched discussions, but not the issues. there's a bunch of conversation about it in the early issues of the repo

**Comment by doozMen** [comment]
I do it in my code. You can however never get compile time safety, runtime you can. The trick is to wrap it in a factory, not just return fatalError, as that will always fail even with a registration as you do not actually create a factory, you do not even return when you try to access the factory of 

```swift
ipublic protocol AuthClientType {
    func isLoggedIn() async -> Bool
    func token() async throws -> String?
}

extension Container {
    var authClient: Factory<AuthClientType> {
       // wrap the fatal error in a factory the somehow weird `self { ... }` 
        self {  fatalError() }
    }
}
```
I found it more convenient to create a generic promised() function for this use case

```swift
/// Although a nice library `Factory` does only provide an optional promised and considers this a benefit.
/// We disagree with that and have to extend to have a required promised.
extension ManagedContainer {
  public func promised<T>(key: String = #function) -> Factory<T> {
    Factory<T>(self, key: key) {
      fatalError("No registration provided for \(key)")
    }
  }
}

```

then you could change your code to
```swift
extension Container {
    var authClient: Factory<AuthClientType> {
      promised()
    }
}
```
